analysis.py

```diff
@@ -1,114 +1,71 @@
-            # however we expect to have a tree here where this should not happen
-            return True
-    return False
-
-
-def append_subtask(task_type, table_rows, subtasks):
-    ids = {t[2]["id"] for t in subtasks}
-    for id in ids:
-        filtered_tasks = []
-        for idx, subtask in enumerate(subtasks):
-            if "id" not in subtask[2]:
-                continue
-            if id == subtask[2]["id"]:
-                filtered_tasks.append(subtask)
-
-        joined_content = "<br>------------------------------------<br>".join(
-            map(
-                lambda t: json.dumps(t, sort_keys=True, indent=4).replace("\n", "<br>"),
-                filtered_tasks,
-            )
-        )
-        start = filtered_tasks[0][0]
-        finish = filtered_tasks[-1][0]
-        table_rows.append(
-            {
-                "id": id,
-                "type": task_type,
-                "duration": calculate_duration(start, finish),
-                "description": joined_content,
-            }
-        )
+LOGGER_DATE_FMT = "%Y-%m-%d %H:%M:%S.%f"
+
 
+class Content(NamedTuple):
+    timestamp: Any
+    event: Any
+    json: Dict[Any, Any]
 
-def calculate_duration(start, finish):
-    start_datetime = datetime.strptime(start, "%Y-%m-%d %H:%M:%S.%f")
-    finish_datetime = datetime.strptime(finish, "%Y-%m-%d %H:%M:%S.%f")
-    return finish_datetime - start_datetime
-
-
-def create_task_brackets(stripped_content):
-    task_indices = []
-    for i, start_content in enumerate(stripped_content):
-        start_event = start_content.event
-        start_id = start_content.json["id"]
-        if start_event.lower() == "starting task":
-            for j, following_content in enumerate(stripped_content[i:], start=i):
-                following_event = following_content.event
-                following_id = following_content.json["id"]
-                if "task" not in following_content.json:
-                    continue
-                ids_identical = start_id == following_id
-                event_matches_successful = following_event.lower() == "task successful"
-                event_matches_errored = following_event.lower() == "task errored"
-                if ids_identical and (event_matches_successful or event_matches_errored):
-                    task_indices.append([i, j])
-                    break
-    return task_indices
-
-
-def read_raw_content(input_file):
-    content = []
-    if not input_file:
-        content = [line.strip() for line in fileinput.input(input_file)]
+
+class CSVRow(NamedTuple):
+    num: int
+    task_type: str
+    duration: float
+    nodes_involved: int
+
+
+def read_raw_content(input_file: str) -> Tuple[List[Content], str]:
+    content: List[str] = []
+
+    if input_file.endswith('gz'):
+        with gzip.open(input_file, "r") as f:
+            content = [line.strip().decode() for line in f.readlines()]
```

Comment:
Why do we need `decode` here?

Reply:
I have not researched it further, but gzips `open()` seems to produce a different encoding from regular `open()`. With the extra `decode()`, the input between the two versions matches.